### PR TITLE
Fw 1559

### PR DESCRIPTION
**Pull request by mhevery**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [X] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

https://angular-team.atlassian.net/browse/FW-1559

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

### CODE REVIEW

File: packages/core/src/render3/node_manipulation.ts [hunk 1]
```diff
@@ -663,27 +663,54 @@ export function appendChild(childEl: RNode | RNode[], childTNode: TNode, current
...
+    } else {
+      // If we have `ElementContainer` or `IcuContainer` than we need to descend into them and
+      // find the actual first `Element`
+      while (tNodeType === TNodeType.ElementContainer || tNodeType === TNodeType.IcuContainer) {
+        tNode = tNode.child !;
+        ngDevMode && assertDefined(tNode, 'ElementContainer and IcuContainer must have children');
+        tNodeType = tNode.type;
       }
-      return getNativeByTNodeOrNull(tViewNodeChild, lView);
+      return getNativeByTNode(tNode, lView);
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/node_manipulation.ts]: I _think_ that it should be recursive in nature - what might happen is that a child of ElementContainer / IcuContainer is a container which might have views etc.

File: packages/core/src/render3/node_manipulation.ts [hunk 2]
```diff
@@ -663,27 +663,54 @@ export function appendChild(childEl: RNode | RNode[], childTNode: TNode, current
...
+    if (tNode !== null) {
+    let tNodeType = tNode.type;
+    if (tNodeType === TNodeType.Container) {
+      // In this case the first `TNode` is another `LContainer` this means we now need to
+      // recurse
+      // into it the `LView` and retrieve its `RNode`.
+      const childLContainer = lView[tNode.index] as LContainer;
+      ngDevMode && assertLContainer(childLContainer);
+      return getFirstRNodeFromLViewInLContainer(childLContainer, 0);
+    } else {
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/node_manipulation.ts]: I don't see the projection handled here, but it might show up at the root of a view, right?

**Review by pkozlowski-opensource — CHANGES_REQUESTED** [review]
@mhevery I did the first, quick pass on this PR and I think that there are more cases to cover here. Essentially the function you are modifying is _almost_ the same as collecting root nodes (https://github.com/angular/angular/blob/ac9d044cad4fee47ed6bfd44f905b546b41952cc/packages/core/src/render3/view_ref.ts#L305-L348) - the way I understand it is that _only_ difference is that we can stop at the _first_ result.

More generally, I think that we can fold all those root node traversal routines we've got into one - we've got already `applyView` - we would have to "just" add a new tree walk action (collect).

I think that, in this PR, we should:
- add more tests / asserts for all the node types (need to cover the same cases as in `collectNativeNodes` linked above);
- add / modify implementation that covers all the cases.

Then, with tests in place and in a separate PR, refactor code to have the root nodes tree walking in one place only (should drop quite a bit of code duplication!). I've spent quite a bit of time poking around the code so happy to discuss this in more details!

**Comment by mhevery** [comment]
@pkozlowski-opensource I have refactored it to use existing code in 9465a3862d, but as you can see the aesthetics of it are not great. :-( My current thinking is that we should refactor `applyView` and friends into a cursor pattern. We could than just iterate over the cursor as many times as we need to to either add/remove or to retrieve the first item from the list. What do you think?

File: packages/core/test/acceptance/view_insertion_spec.ts [hunk 3]
```diff
@@ -249,4 +277,150 @@ describe('view insertion', () => {
       expect(fixture.debugElement.queryAll(By.css('div.dynamic')).length).toBe(4);
     });
   });
+
+  describe('regression', () => {
+    it('should allow inserting of all different types', () => {
```
Comment (pkozlowski-opensource) [inline packages/core/test/acceptance/view_insertion_spec.ts]: Could we split this test into several ones (one test for each case, similarly to what I did in https://github.com/angular/angular/pull/33493? )? It is super-hard to follow-as is.

File: packages/core/test/acceptance/view_insertion_spec.ts [hunk 4]
```diff
@@ -249,4 +277,150 @@ describe('view insertion', () => {
       expect(fixture.debugElement.queryAll(By.css('div.dynamic')).length).toBe(4);
     });
   });
+
+  describe('regression', () => {
```
Comment (pkozlowski-opensource) [inline packages/core/test/acceptance/view_insertion_spec.ts]: nits regarding naming:
* regression -> non-regression? but more generally, I would have sth more descriptive, like "should allow inserting views before views with different node types";
* "should allow inserting of all different types" - the essence of those tests is that we want to insert _before_ a view with specific nodes layout and it should be reflected in the name.

Generally speaking, I would love to have this test split into multiple ones (see below)

**Comment by mhevery** [comment]
@pkozlowski-opensource took over this one here https://github.com/angular/angular/pull/33627
